merge.go

```diff
@@ -0,0 +1,593 @@
+			{
+				node: dbParent,
+			}}
+		dbGraph.NewEdge(dbRoot, dbParent, ls.HasTerm, nil)
+		// dbGraph.NewEdge(dbParent, dbParent, ls.HasTerm, nil)
+	}
+
+	for memItr.Next() {
+		memChildNode := memItr.Edge().GetTo()
+		memChildEdge := memItr.Edge()
+		if dbNode == nil {
+			n := dbGraph.NewNode(memChildNode.GetLabels().Slice(), ls.CloneProperties(memChildNode))
+			e := dbGraph.NewEdge(dbParent, n, memChildEdge.GetLabel(), ls.CloneProperties(memChildEdge))
+			nodeAssociations[memChildNode] = []step{
+				{
+					node: n,
+					edge: e,
+				}}
+			edgeAssociations[memChildEdge] = []edgeStep{
+				{
+					to:   n,
+					from: memChildEdge.GetFrom(),
+					edge: e,
+				}}
+			continue
+		}
+		dbItr := dbNode.GetEdges(lpg.OutgoingEdge)
+		// creates
+		if !dbItr.Next() {
+			//n := dbGraph.NewNode(memNode.GetLabels().Slice(), ls.CloneProperties(memNode))
+			n := dbGraph.NewNode(memChildNode.GetLabels().Slice(), ls.CloneProperties(memChildNode))
+			dbGraph.NewEdge(dbNode, n, memChildEdge.GetLabel(), ls.CloneProperties(memChildEdge))
+			unmapped = append(unmapped, step{node: memChildNode, edge: memChildEdge})
+		}
+		// updates
+		for dbItr.Next() {
+			// dbChildNode := dbItr.Edge().GetTo()
+			dbChildEdge := dbItr.Edge()
+			mpv, ok := memChildNode.GetProperty(ls.SchemaNodeIDTerm)
+			if !ok {
+				break
+			}
+			dbpv, ok := memChildNode.GetProperty(ls.SchemaNodeIDTerm)
+			if !ok {
+				break
+			}
+			if lpg.ComparePropertyValue(mpv, dbpv) == 0 {
+				nodeAssociations[memChildNode] = []step{
+					{
+						node: dbChildEdge.GetFrom(),
+						edge: dbChildEdge,
+					}}
+				edgeAssociations[memChildEdge] = []edgeStep{
+					{
+						to:   dbChildEdge.GetTo(),
+						from: dbChildEdge.GetFrom(),
+						edge: dbChildEdge,
+					}}
+				break
+			}
+		}
+	}
+	for mn, steps := range nodeAssociations {
+		for _, step := range steps {
+			dn := compareGraphNode(mn, step.node, mergeOp)
+			dn.id = dbGraphIds[step.node]
+			dn.operation = updateNodeOp
+			dn.node = step.node
+			dn.setDeltaToNode(step.node)
+			deltas = append(deltas, dn)
+		}
+	}
+	for memEdge, steps := range edgeAssociations {
+		for _, step := range steps {
+			de := compareGraphEdge(memEdge, step.edge, mergeOp)
+			de.id = dbEdges[step.edge]
+			de.operation = updateEdgeOp
+			de.setDeltaToEdge(step.edge)
+			deltas = append(deltas, de)
+		}
+	}
+	for _, step := range unmapped {
+		dn := compareGraphNode(step.node, nil, overwriteOp)
+		dn.operation = createNodeOp
+		dn.id = dbGraphIds[step.node]
+		dn.setDeltaToNode(step.node)
+		dn.node = step.node
+		deltas = append(deltas, dn)
+		de := compareGraphEdge(step.edge, nil, overwriteOp)
+		de.operation = createEdgeOp
+		de.id = dbEdges[step.edge]
+		de.edge = step.edge
+		de.setDeltaToEdge(step.edge)
+		deltas = append(deltas, de)
+	}
+	return deltas
+}
+
+func mergeEntity(memParent, dbParent *lpg.Node, dbGraph *lpg.Graph, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64) ([]delta, bool) {
```

Comment:
This is broken.
Rename memParent -> memEntityRoot, dbParent -> dbEntityRoot

mergeEntity(memEntityRot, dbEntityroot)
calls mergeSubtree(memEntityRoot, dbEntityRoot)

For each child of memParent that is not another entity (i.e. does not have EntitySchema property), find a matching child of dbParent (match schemaNodeIds). If found, merge them. If not found, insert the mem child. Do this recursively

Algorithm sketch for mergeSubtree
For each child of memparent
  if dbroot is nil, insert the node, continue
  if child is another entity, continue
  find a db child with same schemaNodeId
  If found, merge, and do this recursively
  if not found, insert, and do recursively